### PR TITLE
[WIP] Scene Interactor

### PR DESCRIPTION
**Pull request by SPaluchiewiczSofomo**

### CODE REVIEW

File: ios/RNMagicScript/components/BaseNodes/Scene/SceneInteractor.swift [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+            #endif
+        }
+    }
+
+    private func updatePrismMenu(for prism: Prism) {
+        scene?.prismContextMenu.prism = prism
+        var scaleFactor: Float = 1.0
+        scaleFactor = max(scaleFactor, prism.scale.x)
+        scaleFactor = max(scaleFactor, prism.scale.y)
+        scaleFactor = max(scaleFactor, prism.scale.z)
```
Comment (pleszkiewicz) [inline ios/RNMagicScript/components/BaseNodes/Scene/SceneInteractor.swift:99]:
```
// let scaleFactor: Float = max(prism.scale.x, max(prism.scale.y, prism.scale.z)) 

let scaleFactor: Float = prism.scale.y
```

File: ios/RNMagicScript/components/UiNodes/Toggle/UiToggleGroupNode.swift [hunk 2]
```diff
@@ -143,7 +143,7 @@ import SceneKit
             node.on = !node.on
         } else if allowMultipleOn == false && allowAllOff {
             let filteredItems = itemsList.filter { $0.on }
-            if filteredItems.count >= 1 && filteredItems.contains { $0 != node }  { filteredItems.forEach { $0.on = false } }
+            if filteredItems.count >= 1 && filteredItems.contains(where: { $0 != node }) { filteredItems.forEach { $0.on = false } }
```
Comment (pleszkiewicz) [inline ios/RNMagicScript/components/UiNodes/Toggle/UiToggleGroupNode.swift:146]: 👍

File: ios/RNMagicScript/components/BaseNodes/Scene/Scene.swift [hunk 3]
```diff
@@ -26,6 +26,12 @@ import SceneKit
         get { return prisms.filter({ $0.editMode }).first }
         set { }
     }
+    
+    @objc var prismContextMenu: PrismContextMenu! {
```
Comment (pleszkiewicz) [inline ios/RNMagicScript/components/BaseNodes/Scene/Scene.swift:30]: optional?

File: ios/RNMagicScript/components/BaseNodes/Scene/Scene.swift [hunk 4]
```diff
@@ -26,6 +26,12 @@ import SceneKit
...
+        didSet {
+            addChildNode(prismContextMenu)
```
Comment (pleszkiewicz) [inline ios/RNMagicScript/components/BaseNodes/Scene/Scene.swift:32]: Should we remove `prismContextMenu` on nil?

File: ios/RNMagicScript/components/BaseNodes/Scene/SceneInteractor.swift [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+//sourcery: AutoMockable
+protocol SceneInteracting: class {
+    func onUpdate(pointOfView: SCNNode, atTime: TimeInterval)
+}
+
+class SceneInteractor {
+    weak var scene: Scene? {
+        didSet {
+            scene?.prismContextMenu = PrismContextMenuBuilder.build(prismInteractor: prismInteractor)
```
Comment (pleszkiewicz) [inline ios/RNMagicScript/components/BaseNodes/Scene/SceneInteractor.swift:34]: what if `scene` is nil?

File: ios/RNMagicScript/components/BaseNodes/Scene/SceneInteractor.swift [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+                    return distSq1 < distSq2
+                }
+
+                let pointedPrism = results.first!.prism
+                if pointedPrism != lastPointedPrims {
+                    pointedPrism.isPointed = true
+                    prismInteractor.interactedPrism = pointedPrism
+                }
+                updatePrismMenu(for: pointedPrism)
+                DispatchQueue.main.async {
```
Comment (pleszkiewicz) [inline ios/RNMagicScript/components/BaseNodes/Scene/SceneInteractor.swift:75]: I would move it back to NodesManager+Extension.

File: ios/RNMagicScript/components/BaseNodes/Scene/SceneInteractor.swift [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+                scene?.prismContextMenu.isHidden = false
+                results.sort { (item1, item2) -> Bool in
+                    let distSq1 = (item1.intersectionPoint - ray.begin).lengthSq()
+                    let distSq2 = (item2.intersectionPoint - ray.begin).lengthSq()
+                    return distSq1 < distSq2
+                }
+
+                let pointedPrism = results.first!.prism
+                if pointedPrism != lastPointedPrims {
+                    pointedPrism.isPointed = true
```
Comment (pleszkiewicz) [inline ios/RNMagicScript/components/BaseNodes/Scene/SceneInteractor.swift:71]:
```
lastPointedPrims.isPointed = false
```

**Comment by SPaluchiewiczSofomo** [comment]
@pleszkiewicz I'll close this PR - I'm going to integrate this change in bugfix PR.

**Pull request closed without merging**